### PR TITLE
Add prototype for installing apps

### PR DESCRIPTION
**Pull request by alexellis**
Signed-off-by: Alex Ellis (OpenFaaS Ltd) <alexellis2@gmail.com>

## Motivation and Context

As more and more people use k3sup to demo Kubernetes at meetups
and at work etc, it'd be great to provide a 1-click experience of how to
then add OpenFaaS and show the UI / something useful to their
audience / colleagues.

This is just a prototype to generate discussion and there are plans to build
an installer for OpenFaaS already - so this could be a good place to 
experiment. https://github.com/openfaas/faas-cli/issues/612

## Description

Add prototype for installing apps. Installs OpenFaaS.

Optional flag shown for `--loadbalancer`.

Pre-reqs are not checked i.e. helm/tiller.

## How Has This Been Tested?

Tested on a k3s cluster

## Context

This is a prototype for discussion only and values are hard-coded.

### CODE REVIEW

**Comment by gosuri** [comment]
Few K8S demos I made in 2015 that I good success with, it aligns with your one-click motivation:

- [Cluster Life-Cycle Mangement](https://www.youtube.com/watch?v=Z-JSQpScYck)
- [No Downtime Deployments](https://www.youtube.com/watch?v=hruZnd63_IA)

As the founder of AngelHack(at the time largest hackathon), I was fortunate to see some great demos. The best one being Firebase, which I helped launch. A common theme:

1) Attention: Show something jaw-dropping (usually productive)
2) Delight: Let the audience experience it themselves
3) Learning: Walk-thru the steps to recreate it

Depending on the audience,  I would optimize for a minimal number of steps and ideally avoid compilation for non-relevant code.

###  Dependency management

This is a rabbit hole without a package manager, I'm not sure if k8s is stable these days but most custom scripts I wrote ended up in the zombie land.

### `curl | sh`

This gets bad rep but I think it works fine for demos as long it comes for a trusted domain.

### Flag Management

spf13/cobra#413 is a stale issue, they fixed persistence. We use it at [Akash](https://github.com/ovrclk/akash/tree/master/cmd)

### armhf/arm64 etc

I think there is a fix possible with the [manifest-tool](https://github.com/estesp/manifest-tool/), have to look into it.

**Comment by gosuri** [comment]
From a project structure standpoint, layer organization with make files do wonders:

Just opened this up (still WIP): https://github.com/ovrclk/stack

**Comment by alexellis** [comment]
Thank you for sharing your thoughts on this 💯 

re: Flag Management, do you have an example?

**Comment by alexellis** [comment]
I've pushed some changes and helper methods - it effectively does a helm template and means that armhf can be detected and used when needed and that flags could be translated to values in the helm chart.

File: pkg/cmd/apps.go [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+			return err
+		}
+
+		err = installOpenFaaS(kubeConfigPath, lb)
+
+		if err != nil {
+			return err
+		}
+
+		err = kubectl("apply", "-f", "https://raw.githubusercontent.com/openfaas/faas-netes/master/namespaces.yml")
```
Comment (LucasRoesler) [inline pkg/cmd/apps.go]: Is this namespace script needed if you also accept a namespace flag?

File: pkg/cmd/apps.go [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+		if err != nil {
+			return err
+		}
+
+		pass, err := password.Generate(25, 10, 0, false, true)
+		if err != nil {
+			return err
+		}
+
+		err = kubectl("-n", "openfaas", "create", "secret", "generic", "basic-auth", "--from-literal=basic-auth-user=admin", "--from-literal=basic-auth-password='"+pass+"', ")
```
Comment (LucasRoesler) [inline pkg/cmd/apps.go]: This also should accept the value from the namespace flag

File: pkg/cmd/apps.go [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+			return err
+		}
+
+		err = updateHelmRepos()
+
+		if err != nil {
+			return err
+		}
+
+		err = installOpenFaaS(kubeConfigPath, lb)
```
Comment (LucasRoesler) [inline pkg/cmd/apps.go]: This method curls and then executes a script that seems to duplicate some.of the install work, e.g. it installs a basic auth password

File: pkg/cmd/apps.go (hunk 1 above)
Comment (alexellis) [reply pkg/cmd/apps.go]: Good point

File: pkg/cmd/apps.go (hunk 2 above)
Comment (alexellis) [reply pkg/cmd/apps.go]: Yes

File: pkg/cmd/apps.go (hunk 3 above)
Comment (alexellis) [reply pkg/cmd/apps.go]: Good point. That's the old code that should be removed now

File: pkg/cmd/apps.go (hunk 1 above)
Comment (alexellis) [reply pkg/cmd/apps.go]: Ideally we'd template the namespace file or replace text within it.

```
  labels:
    role: openfaas-system
    access: openfaas-system
    istio-injection: enabled
```

^ that or codify these labels

Comment (alexellis) [reply pkg/cmd/apps.go]: .. and call `kubectl create ns`

File: pkg/cmd/apps.go (hunk 3 above)
Comment (alexellis) [reply pkg/cmd/apps.go]: Removed, still worked.

File: pkg/cmd/apps.go [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+	return openfaas
+}
+
+func fetchChart(path, chart string) error {
+	mkErr := os.MkdirAll(path, 0700)
+
+	if mkErr != nil {
+		return mkErr
+	}
+	task := execute.ExecTask{
```
Comment (alexellis) [inline pkg/cmd/apps.go]: From @stefanprodan

```
      --home string                     location of your Helm config. Overrides $HELM_HOME (default "/Users/alex/.helm")
```

**Comment by alexellis** [comment]
I'm going to merge the initial version so that people can play with it.

Follow ups:

- download `helm` when not present
- use temp folder for helm's home folder

**Pull request merged**